File: cli/golem-cli/templates/scala/snapshotting/component-dir/src/main/scala/component_name/CounterWithSnapshotAgent.scala
package component_name

import golem.runtime.annotations.{agentDefinition, description, endpoint, prompt}
import golem.BaseAgent

import scala.concurrent.Future

@agentDefinition(mount = "/snapshot-counters/{name}", snapshotting = "every(1)")
trait CounterWithSnapshotAgent extends BaseAgent {

  class Id(val name: String)

  @prompt("Increase the count by one")
  @description("Increases the count by one and returns the new value")
  @endpoint(method = "POST", path = "/increment")
  def increment(): Future[Int]
}

File: cli/golem-cli/templates/ts/snapshotting/component-dir/src/counter-with-snapshot-agent.ts
import { z } from 'zod';
import { defineAgent, method, http } from '@golemcloud/golem-ts-sdk';

// A counter that opts into snapshotting. The `snapshotting` option declares a
// TYPED state schema — only the schema-declared fields of `this` (here `count`)
// are serialized — plus a policy for WHEN to snapshot (every 5 invocations). On
// recovery the executor restores `count` from the last snapshot and replays the
// oplog tail. This is the declarative fluent replacement for overriding
// `save`/`loadSnapshot` in the decorator SDK.
export const CounterWithSnapshot = defineAgent({
  name: 'CounterWithSnapshot',
  id: { name: z.string() },
  http: http.mount('/snapshot-counters/{name}'),
  snapshotting: { state: z.object({ count: z.number() }), policy: { everyNInvocations: 5 } },
  methods: {
    value: method({ input: {}, returns: z.number() }),
    increment: method({ input: {}, returns: z.number(), http: http.post('/increment') }),
  },
});

export const CounterWithSnapshotImpl = CounterWithSnapshot.implement({
  init: () => ({ count: 0 }),
  methods: {
    value() {
      return this.count;
    },
    increment() {
      this.count += 1;
      return this.count;
    },
  },
});
